# Exploratory Data Analysis

For further research you can check <br>
[ESA’S ANNUAL SPACE ENVIRONMENT REPORT](https://www.sdo.esoc.esa.int/environment_report/Space_Environment_Report_latest.pdf) <br>
[Nasa Space Debris](https://www.nasa.gov/headquarters/library/find/bibliographies/space-debris/)

#### Read Data and Imports

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
df = pd.read_csv('space_decay.csv')
df.head()

CCSDS_OMM_VERS                            COMMENT        CREATION_DATE  \
0               2  GENERATED VIA SPACE-TRACK.ORG API  2021-11-01T06:46:11   
1               2  GENERATED VIA SPACE-TRACK.ORG API  2021-11-01T04:58:37   
2               2  GENERATED VIA SPACE-TRACK.ORG API  2021-11-01T06:26:11   
3               2  GENERATED VIA SPACE-TRACK.ORG API  2021-10-31T18:07:15   
4               2  GENERATED VIA SPACE-TRACK.ORG API  2021-11-01T04:58:37   

  ORIGINATOR      OBJECT_NAME   OBJECT_ID CENTER_NAME REF_FRAME TIME_SYSTEM  \
0    18 SPCS  ARIANE 42P+ DEB   1992-072J       EARTH      TEME         UTC   
1    18 SPCS         SL-8 DEB   1979-028C       EARTH      TEME         UTC   
2    18 SPCS           GSAT 1   2001-015A       EARTH      TEME         UTC   
3    18 SPCS         CZ-4 DEB  1999-057MB       EARTH      TEME         UTC   
4    18 SPCS         CZ-4 DEB  1999-057MC       EARTH      TEME         UTC   

  MEAN_ELEMENT_THEORY  ... RCS_SIZE  COUNTRY_CODE  LAUNCH_DATE   SITE  \
0                SGP4  ...   MEDIUM            FR       1992.0  FRGUI   
1                SGP4  ...    SMALL           CIS       1979.0  PKMTR   
2                SGP4  ...    LARGE           IND       2001.0    SRI   
3                SGP4  ...    SMALL           PRC       1999.0    TSC   
4                SGP4  ...    SMALL           PRC       1999.0    TSC   

   DECAY_DATE     FILE      GP_ID          TLE_LINE0  \
0         NaN  3195178  188614016  0 ARIANE 42P+ DEB   
1         NaN  3194950  188593285         0 SL-8 DEB   
2         NaN  3195026  188609573           0 GSAT 1   
3         NaN  3194431  188556894         0 CZ-4 DEB   
4         NaN  3194950  188592541         0 CZ-4 DEB   

                                           TLE_LINE1  \
0  1 26741U 92072J   21304.94919376  .00000883  0...   
1  1 26743U 79028C   21304.68908982  .00000079  0...   
2  1 26745U 01015A   21305.22411368 -.00000165  0...   
3  1 26754U 99057MB  21304.46625230  .00002265  0...   
4  1 26755U 99057MC  21304.74081807  .00002610  0...   

                                           TLE_LINE2  
0  2 26741   7.7156  90.2410 6528926 243.1216  38...  
1  2 26743  82.9193 299.1120 0030720 158.9093 201...  
2  2 26745  12.1717  16.5368 0237386 250.1248 146...  
3  2 26754  98.4781   8.7205 0060618  37.3771 323...  
4  2 26755  98.4232 122.0724 0062255 345.1605  27...  

[5 rows x 40 columns]

## Explanation of Data

In [4]:
df.describe()

CCSDS_OMM_VERS   MEAN_MOTION  ECCENTRICITY   INCLINATION  \
count         14372.0  14372.000000  14372.000000  14372.000000   
mean              2.0     12.463847      0.067168     74.354208   
std               0.0      4.509305      0.181547     29.626780   
min               2.0      0.051783      0.000005      0.001400   
25%               2.0     13.452144      0.000725     53.055300   
50%               2.0     14.277052      0.003416     86.369200   
75%               2.0     14.963745      0.013505     98.565125   
max               2.0     16.404097      0.897218    144.586200   

       RA_OF_ASC_NODE  ARG_OF_PERICENTER  MEAN_ANOMALY  EPHEMERIS_TYPE  \
count    14372.000000       14372.000000  14372.000000         14372.0   
mean       182.353111         165.977199    191.015595             0.0   
std        116.717713         104.461866    109.873807             0.0   
min          0.020000           0.008200      0.004800             0.0   
25%         71.987350          77.363475     90.001600             0.0   
50%        189.460250         150.483900    206.957400             0.0   
75%        292.484750         258.328550    284.163200             0.0   
max        359.967900         359.989900    359.981900             0.0   

        NORAD_CAT_ID  ELEMENT_SET_NO  ...  MEAN_MOTION_DOT  MEAN_MOTION_DDOT  \
count   14372.000000         14372.0  ...     14372.000000      1.437200e+04   
mean    39558.160381           999.0  ...         0.000049      2.307751e-07   
std      7619.846133             0.0  ...         0.002391      1.485939e-05   
min     26741.000000           999.0  ...        -0.014873     -4.210300e-05   
25%     33965.750000           999.0  ...         0.000001      0.000000e+00   
50%     40165.500000           999.0  ...         0.000010      0.000000e+00   
75%     45173.500000           999.0  ...         0.000032      0.000000e+00   
max    270288.000000           999.0  ...         0.225321      1.654800e-03   

       SEMIMAJOR_AXIS        PERIOD       APOAPSIS      PERIAPSIS  \
count    14372.000000  14372.000000   14372.000000   14372.000000   
mean     10636.185210    223.521557    5721.060178    2795.040245   
std       9580.627227    446.517041   13016.299008    7943.045264   
min       6542.850000     87.783000     183.393000      73.279000   
25%       6956.246750     96.232750     599.876000     546.115000   
50%       7177.549000    100.861000     843.686000     702.433500   
75%       7468.055000    107.046000    1207.919500     831.959750   
max     304049.635000  27808.429000  372136.402000  224547.015000   

        LAUNCH_DATE  DECAY_DATE          FILE         GP_ID  
count  14333.000000         0.0  1.437200e+04  1.437200e+04  
mean    2004.507570         NaN  3.194186e+06  1.885638e+08  
std       13.565227         NaN  3.119117e+03  1.393616e+05  
min     1961.000000         NaN  3.142499e+06  1.860926e+08  
25%     1997.000000         NaN  3.194431e+06  1.885665e+08  
50%     2002.000000         NaN  3.194950e+06  1.885986e+08  
75%     2018.000000         NaN  3.195064e+06  1.886107e+08  
max     2021.000000         NaN  3.195727e+06  1.886250e+08  

[8 rows x 22 columns]

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14372 entries, 0 to 14371
Data columns (total 40 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   CCSDS_OMM_VERS       14372 non-null  int64  
 1   COMMENT              14372 non-null  object 
 2   CREATION_DATE        14372 non-null  object 
 3   ORIGINATOR           14372 non-null  object 
 4   OBJECT_NAME          14372 non-null  object 
 5   OBJECT_ID            14333 non-null  object 
 6   CENTER_NAME          14372 non-null  object 
 7   REF_FRAME            14372 non-null  object 
 8   TIME_SYSTEM          14372 non-null  object 
 9   MEAN_ELEMENT_THEORY  14372 non-null  object 
 10  EPOCH                14372 non-null  object 
 11  MEAN_MOTION          14372 non-null  float64
 12  ECCENTRICITY         14372 non-null  float64
 13  INCLINATION          14372 non-null  float64
 14  RA_OF_ASC_NODE       14372 non-null  float64
 15  ARG_OF_PERICENTER    14372 non-null 

In [6]:
df.isnull().sum()

CCSDS_OMM_VERS             0
COMMENT                    0
CREATION_DATE              0
ORIGINATOR                 0
OBJECT_NAME                0
OBJECT_ID                 39
CENTER_NAME                0
REF_FRAME                  0
TIME_SYSTEM                0
MEAN_ELEMENT_THEORY        0
EPOCH                      0
MEAN_MOTION                0
ECCENTRICITY               0
INCLINATION                0
RA_OF_ASC_NODE             0
ARG_OF_PERICENTER          0
MEAN_ANOMALY               0
EPHEMERIS_TYPE             0
CLASSIFICATION_TYPE        0
NORAD_CAT_ID               0
ELEMENT_SET_NO             0
REV_AT_EPOCH               0
BSTAR                      0
MEAN_MOTION_DOT            0
MEAN_MOTION_DDOT           0
SEMIMAJOR_AXIS             0
PERIOD                     0
APOAPSIS                   0
PERIAPSIS                  0
OBJECT_TYPE                0
RCS_SIZE                 198
COUNTRY_CODE              39
LAUNCH_DATE               39
SITE                      39
DECAY_DATE    

In [7]:
df.columns

Index(['CCSDS_OMM_VERS', 'COMMENT', 'CREATION_DATE', 'ORIGINATOR',
       'OBJECT_NAME', 'OBJECT_ID', 'CENTER_NAME', 'REF_FRAME', 'TIME_SYSTEM',
       'MEAN_ELEMENT_THEORY', 'EPOCH', 'MEAN_MOTION', 'ECCENTRICITY',
       'INCLINATION', 'RA_OF_ASC_NODE', 'ARG_OF_PERICENTER', 'MEAN_ANOMALY',
       'EPHEMERIS_TYPE', 'CLASSIFICATION_TYPE', 'NORAD_CAT_ID',
       'ELEMENT_SET_NO', 'REV_AT_EPOCH', 'BSTAR', 'MEAN_MOTION_DOT',
       'MEAN_MOTION_DDOT', 'SEMIMAJOR_AXIS', 'PERIOD', 'APOAPSIS', 'PERIAPSIS',
       'OBJECT_TYPE', 'RCS_SIZE', 'COUNTRY_CODE', 'LAUNCH_DATE', 'SITE',
       'DECAY_DATE', 'FILE', 'GP_ID', 'TLE_LINE0', 'TLE_LINE1', 'TLE_LINE2'],
      dtype='object')

### Analyzing of Columns

#### Metadata & Identification Columns

CCSDS_OMM_VERS – Version of the CCSDS (Consultative Committee for Space Data Systems) Orbit Mean-Elements Message (OMM) format. ['2'] <br>
COMMENT – Any additional notes or remarks.['GENERATED VIA SPACE-TRACK.ORG API']<br>
CREATION_DATE – Date when the data was generated.<br>
ORIGINATOR – The organization responsible for generating the data.['18 SPCS']<br>
OBJECT_NAME – The name of the object (e.g., a satellite or debris).<br>
OBJECT_ID – The official identifier of the object, typically indicating the launch year, launch number, and piece of the launch (e.g., "2022-045A").<br>
CENTER_NAME – The central body around which the object orbits (e.g., Earth).<br>
REF_FRAME – The reference coordinate frame used (e.g., Earth-Centered Inertial).<br>
TIME_SYSTEM – The time standard used in the dataset (e.g., UTC, TAI).<br>
MEAN_ELEMENT_THEORY – The theoretical model used for mean orbital elements.<br>

>After looking at unique values OBJECT_NAME and OBJECT_ID will be taken to dataset (others have only one unique value)

In [8]:
df = df.drop(['CCSDS_OMM_VERS', 'COMMENT','CREATION_DATE', 'ORIGINATOR', 'CENTER_NAME', 'REF_FRAME', 'TIME_SYSTEM','MEAN_ELEMENT_THEORY'], errors='ignore', axis=1)

#### Orbital Elements

EPOCH – The specific time at which the orbital elements are valid.<br>
MEAN_MOTION – The number of orbits the object completes per day.<br>
ECCENTRICITY – The shape of the orbit, where 0 = circular and values closer to 1 = highly elliptical.<br>
INCLINATION – The tilt of the orbit relative to Earth's equatorial plane, in degrees.<br>
RA_OF_ASC_NODE (Right Ascension of Ascending Node - RAAN) – The longitude of the point where the object crosses the equatorial plane going north.<br>
ARG_OF_PERICENTER – The angle that defines where the closest approach (perigee) happens in the orbit.<br>
MEAN_ANOMALY – The fraction of the orbit completed since the last perigee, expressed in degrees.

>In this part all columns have important information so I will keep all

### Additional Orbital Parameters

EPHEMERIS_TYPE – The type of ephemeris data used (typically 0 for Two-Line Elements, TLE).<br>
CLASSIFICATION_TYPE – Indicates whether the object is publicly available ('U' for unclassified) or classified.<br>
NORAD_CAT_ID – A unique numerical identifier assigned by NORAD (North American Aerospace Defense Command).<br>
ELEMENT_SET_NO – The version number of the orbital element set.[999]<br>
REV_AT_EPOCH – The number of orbits the object has completed since launch at the epoch time.<br>

>Only NORAD_CAT_ID and REV_AT_EPOCH will stay others have only one value

In [9]:
df = df.drop(columns=['EPHEMERIS_TYPE', 'CLASSIFICATION_TYPE','ELEMENT_SET_NO'], errors='ignore', axis=1)

### Drag & Perturbation Parameters

BSTAR – The drag coefficient that represents atmospheric drag effects on the object.<br>
MEAN_MOTION_DOT – The rate of change of the mean motion due to perturbations (e.g., atmospheric drag).<br>
MEAN_MOTION_DDOT – The second derivative of the mean motion, indicating further changes in the orbital period.<br>

>In this part all columns have important information so I will keep all

### Physical & Positional Data

SEMIMAJOR_AXIS – The average distance from the object to the center of its orbit, in kilometers.<br>
PERIOD – The time (in minutes) for the object to complete one orbit.<br>
APOAPSIS – The farthest distance (in kilometers) from the Earth in the orbit.<br>
PERIAPSIS – The closest distance (in kilometers) to the Earth in the orbit.<br>

>In this part all columns have important information so I will keep all

### Object Classification & Size

OBJECT_TYPE – The type of object, such as 'PAYLOAD' (satellite), 'DEBRIS', etc.<br>
RCS_SIZE – The radar cross-section size, indicating how large the object appears to radar sensors (small, medium, large, or unknown).

>In this part all columns have important information so I will keep all

### Launch & Decay Information

COUNTRY_CODE – The country responsible for the object.<br>
LAUNCH_DATE – The date when the object was launched.<br>
SITE – The launch site from which the object was sent into space.<br>
DECAY_DATE – The date when the object re-entered Earth’s atmosphere (currently all values are null, meaning no recorded decay events).<br>

>Only DECAY_DATE will be removed because it's null others will stay

In [10]:
df = df.drop(columns=['DECAY_DATE'], errors='ignore', axis=1)

### File & Data Tracking Information and Two-Line Element Set (TLE) Data


FILE – An internal file reference number.<br>
GP_ID – A unique identifier for each general perturbation set of orbital elements.<br>
TLE_LINE0 – The first line of the TLE data, often containing the object’s name.<br>
TLE_LINE1 – The first standard line of the TLE orbital data.<br>
TLE_LINE2 – The second standard line of the TLE orbital data, containing additional orbital parameters.<br>

In [11]:
df = df.drop(columns=['FILE', 'GP_ID', 'TLE_LINE0', 'TLE_LINE1', 'TLE_LINE2'], errors='ignore', axis=1)

In [12]:
df.head()

OBJECT_NAME   OBJECT_ID                       EPOCH  MEAN_MOTION  \
0  ARIANE 42P+ DEB   1992-072J  2021-10-31T22:46:50.340864     2.921700   
1         SL-8 DEB   1979-028C  2021-10-31T16:32:17.360448    13.754973   
2           GSAT 1   2001-015A  2021-11-01T05:22:43.421952     1.038224   
3         CZ-4 DEB  1999-057MB  2021-10-31T11:11:24.198720    14.775907   
4         CZ-4 DEB  1999-057MC  2021-10-31T17:46:46.681248    14.724482   

   ECCENTRICITY  INCLINATION  RA_OF_ASC_NODE  ARG_OF_PERICENTER  MEAN_ANOMALY  \
0      0.652893       7.7156         90.2410           243.1216       38.7796   
1      0.003072      82.9193        299.1120           158.9093      201.3337   
2      0.023739      12.1717         16.5368           250.1248      146.2900   
3      0.006062      98.4781          8.7205            37.3771      323.1632   
4      0.006226      98.4232        122.0724           345.1605       27.6061   

   NORAD_CAT_ID  ...  MEAN_MOTION_DDOT  SEMIMAJOR_AXIS    PERIOD   APOAPSIS  \
0         26741  ...               0.0       20668.618   492.864  27784.871   
1         26743  ...               0.0        7358.038   104.689   1002.507   
2         26745  ...               0.0       41197.852  1386.984  35797.696   
3         26754  ...               0.0        7015.076    97.456    679.465   
4         26755  ...               0.0        7031.400    97.796    697.039   

   PERIAPSIS  OBJECT_TYPE  RCS_SIZE  COUNTRY_CODE LAUNCH_DATE   SITE  
0    796.095       DEBRIS    MEDIUM            FR      1992.0  FRGUI  
1    957.299       DEBRIS     SMALL           CIS      1979.0  PKMTR  
2  33841.738      PAYLOAD     LARGE           IND      2001.0    SRI  
3    594.417       DEBRIS     SMALL           PRC      1999.0    TSC  
4    609.491       DEBRIS     SMALL           PRC      1999.0    TSC  

[5 rows x 23 columns]

In [13]:
df.columns

Index(['OBJECT_NAME', 'OBJECT_ID', 'EPOCH', 'MEAN_MOTION', 'ECCENTRICITY',
       'INCLINATION', 'RA_OF_ASC_NODE', 'ARG_OF_PERICENTER', 'MEAN_ANOMALY',
       'NORAD_CAT_ID', 'REV_AT_EPOCH', 'BSTAR', 'MEAN_MOTION_DOT',
       'MEAN_MOTION_DDOT', 'SEMIMAJOR_AXIS', 'PERIOD', 'APOAPSIS', 'PERIAPSIS',
       'OBJECT_TYPE', 'RCS_SIZE', 'COUNTRY_CODE', 'LAUNCH_DATE', 'SITE'],
      dtype='object')

## Statistical Analysis

#### Cleaning the Data

In [30]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 14333 entries, 0 to 14332
Data columns (total 23 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   OBJECT_NAME        14333 non-null  object 
 1   OBJECT_ID          14333 non-null  object 
 2   EPOCH              14333 non-null  object 
 3   MEAN_MOTION        14333 non-null  float64
 4   ECCENTRICITY       14333 non-null  float64
 5   INCLINATION        14333 non-null  float64
 6   RA_OF_ASC_NODE     14333 non-null  float64
 7   ARG_OF_PERICENTER  14333 non-null  float64
 8   MEAN_ANOMALY       14333 non-null  float64
 9   NORAD_CAT_ID       14333 non-null  int64  
 10  REV_AT_EPOCH       14333 non-null  int64  
 11  BSTAR              14333 non-null  float64
 12  MEAN_MOTION_DOT    14333 non-null  float64
 13  MEAN_MOTION_DDOT   14333 non-null  float64
 14  SEMIMAJOR_AXIS     14333 non-null  float64
 15  PERIOD             14333 non-null  float64
 16  APOAPSIS           14333 no

In [20]:
df.isna().sum()

OBJECT_NAME            0
OBJECT_ID             39
EPOCH                  0
MEAN_MOTION            0
ECCENTRICITY           0
INCLINATION            0
RA_OF_ASC_NODE         0
ARG_OF_PERICENTER      0
MEAN_ANOMALY           0
NORAD_CAT_ID           0
REV_AT_EPOCH           0
BSTAR                  0
MEAN_MOTION_DOT        0
MEAN_MOTION_DDOT       0
SEMIMAJOR_AXIS         0
PERIOD                 0
APOAPSIS               0
PERIAPSIS              0
OBJECT_TYPE            0
RCS_SIZE             198
COUNTRY_CODE          39
LAUNCH_DATE           39
SITE                  39
dtype: int64

In [21]:
bs= pd.isnull(df["OBJECT_ID"])
data = df[bs]
data.head(5)

OBJECT_NAME OBJECT_ID                       EPOCH  \
14333  TBA - TO BE ASSIGNED       NaN  2021-10-26T19:33:01.154016   
14334  TBA - TO BE ASSIGNED       NaN  2021-10-30T09:10:52.021920   
14335  TBA - TO BE ASSIGNED       NaN  2021-10-25T11:51:13.256352   
14336  TBA - TO BE ASSIGNED       NaN  2021-10-30T04:51:21.536640   
14337  TBA - TO BE ASSIGNED       NaN  2021-10-31T03:16:11.538624   

       MEAN_MOTION  ECCENTRICITY  INCLINATION  RA_OF_ASC_NODE  \
14333    14.072310      0.027466      97.6269        299.8786   
14334    13.712874      0.006261     100.4025        355.8349   
14335    14.201848      0.008714      70.5314        351.9159   
14336    13.754670      0.008960      66.7586        284.9364   
14337    14.363799      0.005253      65.7298        124.7260   

       ARG_OF_PERICENTER  MEAN_ANOMALY  NORAD_CAT_ID  ...  MEAN_MOTION_DDOT  \
14333           122.7192      240.0780         81122  ...               0.0   
14334           113.7771      246.9954         81161  ...               0.0   
14335            49.4088      311.4575         81228  ...               0.0   
14336           166.6714      207.4307         81301  ...               0.0   
14337           230.0421      129.6052         81374  ...               0.0   

       SEMIMAJOR_AXIS   PERIOD  APOAPSIS  PERIAPSIS  OBJECT_TYPE  RCS_SIZE  \
14333        7247.000  102.329  1067.908    669.822          TBA       NaN   
14334        7373.090  105.011  1041.115    948.794          TBA       NaN   
14335        7202.865  101.395   887.498    761.961          TBA       NaN   
14336        7358.146  104.692  1045.939    914.082          TBA       NaN   
14337        7148.621  100.252   808.039    732.934          TBA       NaN   

       COUNTRY_CODE LAUNCH_DATE SITE  
14333           NaN         NaN  NaN  
14334           NaN         NaN  NaN  
14335           NaN         NaN  NaN  
14336           NaN         NaN  NaN  
14337           NaN         NaN  NaN  

[5 rows x 23 columns]

> As you can see above for all the null values of OBJECT_ID some other columns (COUNTRY_CODE,LAUNCH_DATE,SITE) are null too. So I will delete this 39 row.

In [26]:
df = df.dropna(subset=['OBJECT_ID'])
df.head()

OBJECT_NAME   OBJECT_ID                       EPOCH  MEAN_MOTION  \
0  ARIANE 42P+ DEB   1992-072J  2021-10-31T22:46:50.340864     2.921700   
1         SL-8 DEB   1979-028C  2021-10-31T16:32:17.360448    13.754973   
2           GSAT 1   2001-015A  2021-11-01T05:22:43.421952     1.038224   
3         CZ-4 DEB  1999-057MB  2021-10-31T11:11:24.198720    14.775907   
4         CZ-4 DEB  1999-057MC  2021-10-31T17:46:46.681248    14.724482   

   ECCENTRICITY  INCLINATION  RA_OF_ASC_NODE  ARG_OF_PERICENTER  MEAN_ANOMALY  \
0      0.652893       7.7156         90.2410           243.1216       38.7796   
1      0.003072      82.9193        299.1120           158.9093      201.3337   
2      0.023739      12.1717         16.5368           250.1248      146.2900   
3      0.006062      98.4781          8.7205            37.3771      323.1632   
4      0.006226      98.4232        122.0724           345.1605       27.6061   

   NORAD_CAT_ID  ...  MEAN_MOTION_DDOT  SEMIMAJOR_AXIS    PERIOD   APOAPSIS  \
0         26741  ...               0.0       20668.618   492.864  27784.871   
1         26743  ...               0.0        7358.038   104.689   1002.507   
2         26745  ...               0.0       41197.852  1386.984  35797.696   
3         26754  ...               0.0        7015.076    97.456    679.465   
4         26755  ...               0.0        7031.400    97.796    697.039   

   PERIAPSIS  OBJECT_TYPE  RCS_SIZE  COUNTRY_CODE LAUNCH_DATE   SITE  
0    796.095       DEBRIS    MEDIUM            FR      1992.0  FRGUI  
1    957.299       DEBRIS     SMALL           CIS      1979.0  PKMTR  
2  33841.738      PAYLOAD     LARGE           IND      2001.0    SRI  
3    594.417       DEBRIS     SMALL           PRC      1999.0    TSC  
4    609.491       DEBRIS     SMALL           PRC      1999.0    TSC  

[5 rows x 23 columns]

In [27]:
df.isna().sum()

OBJECT_NAME            0
OBJECT_ID              0
EPOCH                  0
MEAN_MOTION            0
ECCENTRICITY           0
INCLINATION            0
RA_OF_ASC_NODE         0
ARG_OF_PERICENTER      0
MEAN_ANOMALY           0
NORAD_CAT_ID           0
REV_AT_EPOCH           0
BSTAR                  0
MEAN_MOTION_DOT        0
MEAN_MOTION_DDOT       0
SEMIMAJOR_AXIS         0
PERIOD                 0
APOAPSIS               0
PERIAPSIS              0
OBJECT_TYPE            0
RCS_SIZE             159
COUNTRY_CODE           0
LAUNCH_DATE            0
SITE                   0
dtype: int64

> After deleting we see some null values at RCS_SIZE and now we will fill this values

In [31]:
df["RCS_SIZE"].value_counts()

RCS_SIZE
SMALL     8346
LARGE     4189
MEDIUM    1639
Name: count, dtype: int64

>For filling these empty values I will use the frequencies of these three value above 